core/templates/dev/head/filters.js

```diff
@@ -347,23 +347,40 @@ oppia.filter('normalizeWhitespacePunctuationAndCase', [function() {
 
 // Note that this filter removes additional new lines or <p><br></p> tags
 // at the end of the string.
-oppia.filter('removeExtraLines', [function() {
+oppia.filter('removeExtraLinesAndSpaces', [function() {
   return function(string) {
     if (!angular.isString(string)) {
       return string;
     }
-    var BLANK_LINES_TEXT = '<p><br></p>';
-    var EMPTY_PARA_TEXT = '<p></p>';
-    while (1) {
-      var endIndex = string.length;
-      var bStr = string.substring(endIndex - BLANK_LINES_TEXT.length, endIndex);
-      var pStr = string.substring(endIndex - EMPTY_PARA_TEXT.length, endIndex);
-      if (bStr === BLANK_LINES_TEXT) {
-        string = string.substring(0, endIndex - BLANK_LINES_TEXT.length);
-      } else if (pStr === EMPTY_PARA_TEXT) {
-        string = string.substring(0, endIndex - EMPTY_PARA_TEXT.length);
-      } else {
-        break;
+    var PARA_START = '<p>';
+    var PARA_CLOSE = '</p>';
+    var SPACE_CHAR = ' ';
+    var SPACE_ENTITY = '&nbsp;';
+    var NEW_LINE = '<br>';
+    // To check if it wrapped around <p> and doesn't hinder with
+    // conversion between HTML and RTE.
+    if (string[0] === '<' && string[1] === 'p' && string[2] === '>') {
```

Comment:
Please see my earlier comment on the reverted PR. Let's not do string manipulation. It would be better to create a jQuery element representing the DOM fragment and do DOM operations on it instead.

Reply:
Oh Cool! I didn't understand it the last time. Thanks @seanlip I'll get it done right away